File: ts-express/src/codegen.ts
import { Pool } from "pg";

const ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
const CODE_LENGTH = 6;
const MAX_ATTEMPTS = 3;

function randomCode(): string {
  let code = "";
  const bytes = new Uint8Array(CODE_LENGTH);
  crypto.getRandomValues(bytes);
  for (const byte of bytes) {
    code += ALPHABET[byte % ALPHABET.length];
  }
  return code;
}

export async function generateCode(pool: Pool): Promise<string> {
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const code = randomCode();
    const result = await pool.query("SELECT 1 FROM links WHERE code = $1", [code]);
    if (result.rowCount === 0) {
      return code;
    }
  }
  throw new Error("Failed to generate a unique code after multiple attempts");
}

File: go-nethttp/codegen.go
package main

import (
	"context"
	"crypto/rand"
	"errors"
	"math/big"

	"github.com/jackc/pgx/v5/pgxpool"
)

const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
const codeLength = 6
const maxAttempts = 3

func generateCode(ctx context.Context, db *pgxpool.Pool) (string, error) {
	for range maxAttempts {
		code, err := randomCode()
		if err != nil {
			return "", err
		}
		var exists bool
		if err := db.QueryRow(ctx, "SELECT EXISTS(SELECT 1 FROM links WHERE code = $1)", code).Scan(&exists); err != nil {
			return "", err
		}
		if !exists {
			return code, nil
		}
	}
	return "", errors.New("failed to generate unique code after max attempts")
}

func randomCode() (string, error) {
	b := make([]byte, codeLength)
	for i := range b {
		n, err := rand.Int(rand.Reader, big.NewInt(int64(len(alphabet))))
		if err != nil {
			return "", err
		}
		b[i] = alphabet[n.Int64()]
	}
	return string(b), nil
}
